pkg/util/sliding_window.go

```diff
@@ -0,0 +1,218 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package util
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// PollingFunc is polled at specified intervals to collect the value
+// that should be included in the slidingWindow data
+type PollingFunc func() (value float64)
+
+// CallbackFunc is invoked when the polling function collects data and
+// we need to use the newly-calculated values
+type CallbackFunc func(SlidingWindow)
+
+// slidingWindow is an object that polls for a value every `pollingInterval`
+// and then keeps those values for the duration of the `windowSize` that can
+// then be analyzed/averaged/aggregated. This object should be instantiated
+// via `NewSlidingWindow`.
+type slidingWindow struct {
+	bucketIdx       int
+	buckets         []float64
+	bucketsLock     sync.RWMutex
+	callbackFunc    CallbackFunc
+	initialized     bool
+	numBuckets      int
+	numBucketsUsed  int
+	pollingFunc     PollingFunc
+	pollingInterval time.Duration
+	stopChan        chan struct{}
+	stopped         bool
+	ticker          *time.Ticker
+	windowLock      sync.RWMutex
+	windowSize      time.Duration
+}
+
+// SlidingWindow is the public API that we expose from the slidingWindow object
+type SlidingWindow interface {
+	Start(PollingFunc, CallbackFunc) error
+	Stop()
+
+	Average() float64
+	WindowSize() time.Duration
```

Comment:
It'd be good to have a doc comment on each method in the interface (in any interface, I think..)